### PR TITLE
Added register and login routes

### PR DESCRIPTION
**Pull request by mustafahalilovic**

### CODE REVIEW

File: src/repositories/LoginRegisterRepoInterface.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { QueryResult } from 'pg';
+
+export default interface LoginRegisterRepoInterface{
```
Comment (ohmarinseries) [inline src/repositories/LoginRegisterRepoInterface.ts:3]: This is so well written and thought through, well done bro!!

File: src/repositories/loginRegister.repository.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { QueryResult } from "pg";
+import { pool } from "../database/connection";
+import LoginRegisterRepoInterface from "./LoginRegisterRepoInterface";
+
+export default class LoginRegisterRepository implements LoginRegisterRepoInterface{
+
+    async checkEmail (email: string): Promise<QueryResult> {
+        return await pool.query("SELECT * FROM users WHERE email=$1", [email]);
```
Comment (ohmarinseries) [inline src/repositories/loginRegister.repository.ts:8]: I am starting to be an asshole just to warn you, `SELECT *` is second worst query in the world, first one is query with `LIKE` operator in `WHERE` clause. Always write columns that you need, even if you need all of them.

File: src/utils/errorHandler.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
+interface tryCatchError {
+    code: number,
```
Comment (ohmarinseries) [inline src/utils/errorHandler.ts:3]: Nice!

File: src/utils/validateData.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import validator from "validator";
+
+interface ValidateError {
+    valid: Boolean;
+    description: string;
+};
+
+export const validateData = (
+    email: string,
```
Comment (ohmarinseries) [inline src/utils/validateData.ts:9]: This is really not looking good with chaining ifs, but i think in TS its a best you can do ....

File: src/routes/user.route.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import express from 'express';
+import { loginUser } from '../controllers/user.login';
+import { registerUser } from '../controllers/user.register'; 
+
+const userRoutes = express.Router();
+
+userRoutes.post('/users/register', registerUser);
```
Comment (ohmarinseries) [inline src/routes/user.route.ts:7]: I recommend to prefix all endpoints with `/api/v1/`, you don't have to but its good practice.

File: src/migrations/3_create-table-articles.sql [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    picture VARCHAR(100) NOT NULL,
+    description VARCHAR(255),
+    properties JSON,
+    CONSTRAINT fk_type_id
+        FOREIGN KEY(type_id)
+        REFERENCES types(id),
+    CONSTRAINT fk_user_id
+        FOREIGN KEY(user_id)
+        REFERENCES users(id) 
+);
```
Comment (ohmarinseries) [inline src/migrations/3_create-table-articles.sql:19]: Research about database indexes, and create them next time. Its maybe advanced topic, but I believe you are capable of doing it.

File: src/controllers/user.register.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    }:User = req.body;
+
+    try {
+
+        const isValid = validateData(email, password);
+        if(!isValid.valid){
+            return res.status(400).send(isValid);
+        };
+        
+        const register = new LoginRegisterRepository();
```
Comment (ohmarinseries) [inline src/controllers/user.register.ts:27]: Controller calling Repository ??? Controller -> Service -> Repository

File: src/controllers/user.login.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+export const loginUser = async (req: Request, res: Response) => {
+
+    const {
+        email,
+        password
+    } = req.body;
+
+    try {
+        
+        const login = new LoginRegisterRepository();
```
Comment (ohmarinseries) [inline src/controllers/user.login.ts:16]: Again same here

File: src/controllers/user.login.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import { Request, Response } from "express"
+import LoginRegisterRepository from "../repositories/loginRegister.repository";
+import { errorHandler } from "../utils/errorHandler";
+import bcrypt from 'bcrypt';
+import { signJwt } from "../utils/authentication";
+
+export const loginUser = async (req: Request, res: Response) => {
```
Comment (ohmarinseries) [inline src/controllers/user.login.ts:7]: And I think you don't have to create new file for EVERY endpoint, you just need to group them by topic. Login and Register are user endpoints and can be grouped in single controller file. Like you did with repository.

File: package.json [hunk 10]
```diff
@@ -28,12 +28,16 @@
     "typescript": "5.0.2"
   },
   "dependencies": {
+    "@types/pg": "^8.6.6",
+    "@types/validator": "^13.7.14",
```
Comment (ohmarinseries) [inline package.json:32]: XD

File: src/migrations/3_create-table-articles.sql [hunk 11]
```diff
@@ -0,0 +1,19 @@
+
+CREATE TABLE articles(
+    id SERIAL NOT NULL PRIMARY KEY,
+    type_id INT,
+    user_id INT,
```
Comment (ohmarinseries) [inline src/migrations/3_create-table-articles.sql:5]: Soo it these can be NULL???

**Review by ohmarinseries — APPROVED** [review]

**Comment by ohmarinseries** [comment]
Read my comments, fix them and merge it. Improvements are visible, you are getting better everyday.

File: package.json (hunk 10 above)
Comment (mustafahalilovic) [reply package.json:32]: understand 🤣

File: src/controllers/user.register.ts (hunk 7 above)
Comment (mustafahalilovic) [reply src/controllers/user.register.ts:27]: I only use class LoginRegisterRepository in controller to call their methods, and they are for database interaction. Do I need to make service folder for services, like one file for login and register and than call repo functions there?

File: src/migrations/3_create-table-articles.sql (hunk 6 above)
Comment (mustafahalilovic) [reply src/migrations/3_create-table-articles.sql:19]: Okey

File: src/repositories/LoginRegisterRepoInterface.ts (hunk 1 above)
Comment (mustafahalilovic) [reply src/repositories/LoginRegisterRepoInterface.ts:3]: I knoe it is big but it's descriptive 🤣

Comment (mustafahalilovic) [reply src/repositories/LoginRegisterRepoInterface.ts:3]: know*

File: src/repositories/loginRegister.repository.ts (hunk 2 above)
Comment (mustafahalilovic) [reply src/repositories/loginRegister.repository.ts:8]: You are right there, it's not good practice

File: src/routes/user.route.ts (hunk 5 above)
Comment (mustafahalilovic) [reply src/routes/user.route.ts:7]: Okey

File: src/utils/errorHandler.ts (hunk 3 above)
Comment (mustafahalilovic) [reply src/utils/errorHandler.ts:3]: Thanks 😃

File: src/controllers/user.register.ts (hunk 7 above)
Comment (ohmarinseries) [reply src/controllers/user.register.ts:27]: Yup service will proxy repo methods if there is no extra logic. Its design mate

**Comment by mustafahalilovic** [comment]
Thanks bro

**Pull request merged**